src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -0,0 +1,424 @@
+import BasicMAPF.Instances.Agent;
+import BasicMAPF.Instances.MAPF_Instance;
+import BasicMAPF.Instances.Maps.I_Location;
+import BasicMAPF.Solvers.AStar.CostsAndHeuristics.DistanceTableAStarHeuristic;
+import BasicMAPF.Solvers.A_Solver;
+import BasicMAPF.Solvers.ConstraintsAndConflicts.Constraint.ConstraintSet;
+import Environment.Metrics.InstanceReport;
+import TransientMAPF.TransientMAPFSolution;
+import java.util.*;
+
+public class PIBT_Solver extends A_Solver {
+
+    /**
+     * list contains all not handled agents at timestamp t
+     * initialize in each timestamp to A (all agents that not reached their goal yet)
+     * agent in list will be sorted in descending order by priority
+     */
+    private List<Agent> unhandledAgents;
+
+    /**
+     * list contains all taken nodes - nodes required by an agent in the next timeStamp
+     * hence agent can't move to a node in this list
+     * when an agent choose node to move to in next timestamp, the node is added to this list
+     */
+    private List<I_Location> takenNodes;
+
+    /**
+     * Map saving current location for each agent
+     */
+    private HashMap<Agent, I_Location> locations;
+
+//    /**
+//     * Map saving current location for each agent in the specific time stamp t
+//     * in the end of the time stamp, update locations
+//     */
+//    private HashMap<Agent, I_Location> desiredLocationsInNextT;
+
+    /**
+     * Map saving priority of each agent
+     */
+    private HashMap<Agent, Double> priorities;
+
+    /**
+     * heuristic to use in the low level search to find the closets nodes to an agent's goal
+     */
+    private DistanceTableAStarHeuristic heuristic;
+
+
+    /**
+     * object Solution represents the final solution that the algorithm returns
+     */
+    private Solution solution;
+
+    /**
+     * HashMap saves for each agent his plan
+     * this hash built iteratively in every time stamp
+     * in the end of the algorithm this HashMap represent the final solution
+     */
+    private HashMap<Agent, SingleAgentPlan> agentPlans;
+    private int timeStamp;
+
+    /**
+     * The cost function to evaluate solutions with.
+     */
+    private final I_SolutionCostFunction solutionCostFunction;
+    private ConstraintSet constraints;
+
+    /**
+     * constructor
+     */
+    public PIBT_Solver(I_SolutionCostFunction solutionCostFunction) {
+        super.name = "PIBT";
+        this.solutionCostFunction = Objects.requireNonNullElse(solutionCostFunction, new SOCCostFunction());
+    }
+
+    /**
+     * initialize variables relevant to solve MAPF problem using PIBT
+     * @param instance an instance that we are about to solve.
+     * @param parameters parameters for this coming run.
+     */
+    @Override
+    protected void init(MAPF_Instance instance, RunParameters parameters) {
+        super.init(instance, parameters);
+        this.constraints = parameters.constraints == null ? new ConstraintSet(): parameters.constraints;
```

Comment:
simply throw this error when constraint isn't null?

Reply:
yeah

Reply:
I think i don't undestand what you are meaning, since this error crashes the code if there is a constraint (of any kind)

Reply:
Since we ignore some constraints, keeping the code as-is could cause unexpected and hard to track down problems down the line. It is better to fail early and throw an exception if we get any constraints. In other words, we shouldn't pretend to support constraints when we don't.
Maybe a less restrictive way to do this would be to throw an exception only when we end up with a situation we don't support - which is when we found a solution, but some of the agents can't stay where they are forever because of later constraints. This way we can still support constraints that we know how to handle, but if we get to a situation where we ignore constraints we got, we will fail in a clear way rather then produce unexpected results. You can also add a unit test that produces such a situation and asserts that the exception is thrown.
A third way would be to think together of a solution for these constraints and then we can just fully support constraints :)